### PR TITLE
[Feature] Add Container Logs Config to Project's TH Config 

### LINKED ISSUE #1091: [Feature] Move ENABLE_CONTAINER_LOGS Toggle to THConfig
## Feature description

`ENABLE_CONTAINER_LOGS` (backend `Settings.ENABLE_CONTAINER_LOGS`, `backend/app/core/config.py`) is currently the only remaining Test-Harness-specific toggle still controlled purely by an environment variable in `.env`, requiring a full backend container rebuild/restart to change. It logs container lifecycle operations (create, destroy, `docker cp` file transfers) for debugging.

Its sibling flag, `ENABLE_REALTIME_PYTHON_TEST_LOGS`, was migrated in #1038 to a per-project override via a new `th_config` object in the project's JSON config (`th_config.enable_realtime_python_test_logs`), with the env var kept as the instance-wide fallback default. `ENABLE_CONTAINER_LOGS` was intentionally left out of that change and deserves its own issue, since it isn't a drop-in mirror of that pattern.

## Motivation
  For consistency, it would be good to let `ENABLE_CONTAINER_LOGS` be overridden per-project the same way, so operators can enable container-op debugging for a specific project's runs without touching `.env` / rebuilding the backend.

## Proposed change
  1. Add `enable_container_logs: Optional[bool] = None` to `THConfig` (`backend/app/schemas/test_environment_config.py`), matching the `None = defer to env var` convention used by `enable_realtime_python_test_logs`.
  2. Thread an explicit `enable_container_logs: Optional[bool] = None` parameter through the actual log-emitting methods (`SDKContainer.send_command()`; `ContainerManager.create_container()` / `destroy()` / `copy_file_from_container()` /
  `copy_file_to_container()`), falling back to `settings.ENABLE_CONTAINER_LOGS` when `None`.
  3. Update the `.config`-aware call sites that own an `SDKContainer` (`PythonTestCase`, `PythonTestSuite`, `ChipSuite`/`ChipServer`) to resolve `th_config.enable_container_logs` and pass it down.
  4. Leave callers with no config access (`otbr_manager.py`, test-discovery utilities) passing nothing — they keep today's env-var-only behavior, so this is purely additive.
  5. Add unit test coverage for both the new resolution logic and (since none exists today) the underlying `ENABLE_CONTAINER_LOGS` behavior itself.

## Scope note
This was evaluated for inclusion in #1038 but deliberately deferred: it touches ~8-10 call sites across container lifecycle infrastructure rather than a single class, and doing it properly means adding test coverage where there currently is none — a larger, higher-risk change than the rest of that PR.

### Related
  - #1038 (introduced `th_config` and migrated `ENABLE_REALTIME_PYTHON_TEST_LOGS`)
    - [Backend PR#356](https://github.com/project-chip/certification-tool-backend/pull/356) (Observed by a review feedback)

### Use Cases

[Use Case 1]
- Enable containers logs without rebuilding/restarting the TH backend container.
- Verify the logs works as configured.

## Why this isn't a simple mirror of the `th_config.enable_realtime_python_test_logs` change

`ENABLE_REALTIME_PYTHON_TEST_LOGS` is read inside `PythonTestCase` (`test_collections/matter/sdk_tests/support/python_testing/models/test_case.py`), which already exposes `.config` (via `TestCase`), so resolving `th_config` there was a self-contained, ~4-call-site change.

`ENABLE_CONTAINER_LOGS`, by contrast, is read inside two **singleton** infrastructure classes with no project-config access at all:
  - `ContainerManager` (`app/container_manager/container_manager.py`, `metaclass=Singleton`, instantiated once at module import) — 6 log call sites across `create_container()`, `destroy()`, `copy_file_from_container()`, `copy_file_to_container()`.
  - `SDKContainer` (`test_collections/matter/sdk_tests/support/sdk_container.py`, also `metaclass=Singleton`) — 1 log call site in `send_command()`.

Neither class's constructor takes a project/config reference, and `ChipServer` (which wraps `SDKContainer` and is itself a bare singleton) has no `.config` either. The classes that *do* have `.config` (`PythonTestCase`, `PythonTestSuite`, `ChipSuite`/`ChipTest`) sit one or two layers above where the actual `if settings.ENABLE_CONTAINER_LOGS:` checks live.

There is also currently **no unit test coverage at all** for `ENABLE_CONTAINER_LOGS`.

### PR DESCRIPTION
**Pull request by antonio-amjr**
Fix: https://github.com/project-chip/certification-tool/issues/1091
Depends on: https://github.com/project-chip/matter-test-scripts/pull/805
---

### Description
Extends the th_config pattern (introduced in #1038 for prompt timeout and real-time Python logs) to **ENABLE_CONTAINER_LOGS**, so container-operation debug logging can be toggled per-project without editing .env/restarting the backend, while keeping the env var as the instance-wide fallback default.

Unlike the other two th_config flags, this one is checked inside singleton infra classes (**ContainerManager**, **SDKContainer**) with no project-config access — so the override is resolved by .config-aware callers and threaded down as an explicit parameter through the whole call chain: **ChipSuite**/**PythonTestCase**/**PythonTestSuite**/**PayloadParsingTestBaseClass** → **ChipServer**/**MatterYAMLRunner** → **SDKContainer** → **ContainerManager**.
Callers with no config access (otbr_manager.py, test-discovery utilities) are untouched — they keep the env-var-only default.

### Changed
- `schemas/test_environment_config.py`: THConfig.enable_container_logs: Optional[bool] = None; added `get_th_config_value()` helper that tolerates th_config being a plain dict (production shape) or a pydantic object (some test doubles fake .config via Model.__dict__, which doesn't recursively convert nested models).
- `container_manager/container_manager.py`: new resolve_container_logs_enabled(); enable_container_logs param threaded through create_container/destroy/copy_file_from_container/copy_file_to_container.
- sdk_container.py, chip/chip_server.py, yaml_tests/matter_yaml_runner.py: same param threaded through start/stop/send_command/copy_file_*/generate_manual_pairing_code_with_chip_tool/restart.
- **ChipSuite**, **PythonTestCase**, **PythonTestSuite**, **PayloadParsingTestBaseClass** (nested submodule): each gets a `_container_logs_enabled()` resolver, plus a `_safe_config()` guard so it falls back to the env var instead of raising when .config itself is unavailable (e.g. test doubles with no wired-up project chain).
- default_project.config: added th_config.enable_container_logs example.
- Full new unit test coverage for `resolve_container_logs_enabled()` and the emit/suppress behavior of every touched method (previously zero coverage existed for ENABLE_CONTAINER_LOGS at all).

### CODE REVIEW

**Comment by antonio-amjr** [comment]
@coderabbitai full review

File: app/container_manager/container_manager.py [hunk 1]
```diff
@@ -42,21 +42,42 @@
...
+        self,
+        docker_image_tag: str,
+        parameters: Dict = {},
+        enable_container_logs: Optional[bool] = None,
     ) -> Container:
-        container = self.__run_new_container(docker_image_tag, parameters)
+        logs_enabled = resolve_container_logs_enabled(enable_container_logs)
+        container = self.__run_new_container(
+            docker_image_tag, parameters, logs_enabled
+        )
```
[the remark this answers is not in the document]
Comment (antonio-amjr) [reply app/container_manager/container_manager.py]: Already done.

File: test_collections/matter/sdk_tests/support/python_testing/models/test_suite.py [hunk 2]
```diff
@@ -132,7 +160,9 @@ async def cleanup(self) -> None:
                 logger.warning(f"Could not capture admin_storage.json snapshot: {e}")
 
         logger.info("Stopping SDK container")
-        self.sdk_container.destroy()
+        self.sdk_container.destroy(
+            enable_container_logs=self._container_logs_enabled()
+        )
```
[the remark this answers is not in the document]
Comment (antonio-amjr) [reply test_collections/matter/sdk_tests/support/python_testing/models/test_suite.py]: Already done.

File: test_collections/matter/sdk_tests/support/tests/python_tests/test_python_test_suite.py [hunk 3]
```diff
@@ -618,3 +618,58 @@ async def test_should_perform_new_commissioning_no() -> None:
...
+    [
+        (True, False, True),  # config override True beats env False
+        (False, True, False),  # config override False beats env True
+        (None, True, True),  # unset config defers to env True
+        (None, False, False),  # unset config defers to env False
+    ],
+)
+def test_container_logs_enabled_matrix(
+    th_config_value, env_value, expected
+) -> None:
```
[the remark this answers is not in the document]
Comment (antonio-amjr) [reply test_collections/matter/sdk_tests/support/tests/python_tests/test_python_test_suite.py:634]: Already Done

File: test_collections/matter/sdk_tests/support/yaml_tests/models/chip_suite.py [hunk 4]
```diff
@@ -283,10 +308,12 @@ async def cleanup(self) -> None:
...
 
         logger.info("Stopping test runner")
-        await self.runner.stop()
+        await self.runner.stop(enable_container_logs=self._container_logs_enabled())
 
         logger.info("Stopping SDK container")
-        self.sdk_container.destroy()
+        self.sdk_container.destroy(
+            enable_container_logs=self._container_logs_enabled()
+        )
```
[the remark this answers is not in the document]
Comment (antonio-amjr) [reply test_collections/matter/sdk_tests/support/yaml_tests/models/chip_suite.py]: Done Already

File: test_collections/matter/sdk_tests/support/tests/test_sdk_container.py [hunk 5]
```diff
@@ -171,6 +171,65 @@ async def test_send_command_default_prefix(real_sdk_container) -> None:  # noqa
     real_sdk_container._SDKContainer__container = None
 
 
+@pytest.mark.asyncio
+async def test_send_command_logs_shell_command_when_enabled(
+    real_sdk_container,
```
[the remark this answers is not in the document]
Comment (antonio-amjr) [reply test_collections/matter/sdk_tests/support/tests/test_sdk_container.py]: Moved "# noqa" to the appropriate line.

File: app/tests/container_manager/test_container_manager.py [hunk 6]
```diff
@@ -13,19 +13,89 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+import importlib
 from asyncio import TimeoutError
```
[the remark this answers is not in the document]
Comment (antonio-amjr) [reply app/tests/container_manager/test_container_manager.py]: Done in https://github.com/project-chip/certification-tool-backend/pull/366/commits/492732cd5f5e5f9cf1311c45177a8d70c7e117d0

File: app/core/config.py [hunk 7]
```diff
@@ -111,7 +111,10 @@ def get_emails_enabled(cls, v: bool, values: Dict[str, Any]) -> bool:
     ENABLE_REALTIME_PYTHON_TEST_LOGS: bool = False
 
     # Container Logging
-    ENABLE_CONTAINER_LOGS: bool = False
+    # Defaults to True: most existing TH installs already run with this enabled
+    # via .env, and it can still be explicitly disabled per-instance (.env) or
+    # per-project (th_config.enable_container_logs).
+    ENABLE_CONTAINER_LOGS: bool = True
```
Comment (rquidute) [inline app/core/config.py]: Quick question on this — was flipping the default to True intentional/needed for this PR, or just a side change? Container logging is pretty verbose (shell-equivalent docker run/stop/kill/rm commands, file copy details, etc.), so this would add a fair bit of noise to logs for any install/CI setup that isn't already explicitly setting ENABLE_CONTAINER_LOGS in .env.

Comment (antonio-amjr) [reply app/core/config.py]: Actually, I thought that the default value when installing TH was **True**. So .env comes without it by default?
I'll rollback this then.

Comment (antonio-amjr) [reply app/core/config.py]: Done: https://github.com/project-chip/certification-tool-backend/pull/366/commits/f94fd3beccdbe280a93efaf908dc2b920af64537

**Review by rquidute — APPROVED** [review]

**Comment by oxesoft** [comment]
Reviewed this together with the dependent matter-test-scripts#805 (which looks correct — all 5 `sdk_container` call sites in `PayloadParsingTestBaseClass` are threaded through).

For this PR, I checked out the branch and grepped every non-test call site of `sdk_container.start/destroy/send_command/copy_file_*`, `container_manager.create_container/destroy/copy_file_*`, and `chip_server.start/stop/restart` to check for missed threading of `enable_container_logs`. Found one:

**`should_perform_new_commissioning()` in `test_collections/matter/sdk_tests/support/python_testing/models/utils.py` (~line 351)**

```python
async def should_perform_new_commissioning(
    prompt_support: UserPromptSupport,
    config: TestEnvironmentConfigMatter,
    logger: loguru.Logger,
) -> bool:
    ...
        sdk_container.copy_file_to_container(
            host_file_path=ADMIN_STORAGE_FILE_HOST,
            destination_container_path=storage_path,
        )
```

This is missing `enable_container_logs=_container_logs_enabled(config)`. It's the odd one out — the two sibling functions in this same file, `__copy_admin_storage_file()` (copy *from* container) and `commission_device()`, both correctly pass `enable_container_logs=_container_logs_enabled(config)` using the module-level `_container_logs_enabled()` helper defined right above. `config` is available here too (same `TestEnvironmentConfigMatter` param), so this doesn't look like one of the documented "no config access" exceptions (`otbr_manager.py`, test-discovery utilities) — just a spot missed in the mechanical thread-through.

Effect: when a user resumes a previous commissioning session (admin_storage.json reuse prompt → "yes"), copying that file *into* the container will always follow the `ENABLE_CONTAINER_LOGS` env var and ignore a project's `th_config.enable_container_logs` override, unlike every other file-copy/command path touched by this PR.

Also worth noting: `should_perform_new_commissioning` is mocked out wherever it's called from (`test_python_test_suite.py`), and there's no direct test of its `copy_file_to_container` call in `test_utils.py` — unlike `commission_device`/`__copy_admin_storage_file`, which do have dedicated enabled/disabled tests — so this gap wouldn't be caught by the current suite.

**Comment by antonio-amjr** [comment]
> Reviewed this together with the dependent matter-test-scripts#805 (which looks correct — all 5 `sdk_container` call sites in `PayloadParsingTestBaseClass` are threaded through).
> 
> For this PR, I checked out the branch and grepped every non-test call site of `sdk_container.start/destroy/send_command/copy_file_*`, `container_manager.create_container/destroy/copy_file_*`, and `chip_server.start/stop/restart` to check for missed threading of `enable_container_logs`. Found one:
> 
> **`should_perform_new_commissioning()` in `test_collections/matter/sdk_tests/support/python_testing/models/utils.py` (~line 351)**
> 
> ```python
> async def should_perform_new_commissioning(
>     prompt_support: UserPromptSupport,
>     config: TestEnvironmentConfigMatter,
>     logger: loguru.Logger,
> ) -> bool:
>     ...
>         sdk_container.copy_file_to_container(
>             host_file_path=ADMIN_STORAGE_FILE_HOST,
>             destination_container_path=storage_path,
>         )
> ```
> 
> This is missing `enable_container_logs=_container_logs_enabled(config)`. It's the odd one out — the two sibling functions in this same file, `__copy_admin_storage_file()` (copy _from_ container) and `commission_device()`, both correctly pass `enable_container_logs=_container_logs_enabled(config)` using the module-level `_container_logs_enabled()` helper defined right above. `config` is available here too (same `TestEnvironmentConfigMatter` param), so this doesn't look like one of the documented "no config access" exceptions (`otbr_manager.py`, test-discovery utilities) — just a spot missed in the mechanical thread-through.
> 
> Effect: when a user resumes a previous commissioning session (admin_storage.json reuse prompt → "yes"), copying that file _into_ the container will always follow the `ENABLE_CONTAINER_LOGS` env var and ignore a project's `th_config.enable_container_logs` override, unlike every other file-copy/command path touched by this PR.
> 
> Also worth noting: `should_perform_new_commissioning` is mocked out wherever it's called from (`test_python_test_suite.py`), and there's no direct test of its `copy_file_to_container` call in `test_utils.py` — unlike `commission_device`/`__copy_admin_storage_file`, which do have dedicated enabled/disabled tests — so this gap wouldn't be caught by the current suite.

Done in: https://github.com/project-chip/certification-tool-backend/pull/366/commits/a8885d4cd7bb20bdcce3f9743dfbacc7cd59c7b8

**Review by oxesoft — APPROVED** [review]
